ouroboros-network/src/Ouroboros/Network/Protocol/VersionNegotiation/Test.hs

```diff
@@ -0,0 +1,450 @@
+        prop_arbitrary_ArbitraryValidVersions
+    , testProperty "shrink ArbitraryValidVersions"
+        prop_shrink_ArbitraryValidVersions
+    ]
+  ]
+
+--
+-- Test Versions
+--
+-- Notes: Associated data are choosen in such a way that a decoder will fail
+-- interpreting one of them as the other.  This is done on purpose for testing
+-- missencoded data (protocol version & associated version data mismatch)
+--
+
+-- |
+-- Testing version number
+--
+data VersionNumber
+  = Version_0
+  | Version_1
+  | Version_2
+  deriving (Eq, Ord, Enum, Bounded, Show)
+
+instance Arbitrary VersionNumber where
+  arbitrary = oneof (map return [minBound .. maxBound])
+
+instance Serialise VersionNumber where
+  encode Version_0 = CBOR.encodeWord 0
+  encode Version_1 = CBOR.encodeWord 1
+  encode Version_2 = CBOR.encodeWord 2
+
+  decode = do
+    x <- CBOR.decodeWord
+    case x of
+      0 -> return Version_0
+      1 -> return Version_1
+      2 -> return Version_2
+      _ -> fail "decode VersionNumber: wrong tag"
+
+-- |
+-- Data Associated with @'Version_0'@
+--
+data Data_0 = C0 | C1 | C2
+  deriving (Eq, Show, Typeable, Generic)
+
+instance SerialiseTerm Data_0 where
+  -- We are using @CBOR.TInt@ instead of @CBOR.TInteger@, since for small
+  -- integers generated by QuickCheck they will be encoded as @TkInt@ and then
+  -- are decodec back to @CBOR.TInt@ rather than @COBR.TInteger@.  The same for
+  -- other @SerialiseTerm@ instances in this module.
+  encodeTerm C0 = CBOR.TInt 0
+  encodeTerm C1 = CBOR.TInt 1
+  encodeTerm C2 = CBOR.TInt 2
+
+  decodeTerm (CBOR.TInt 0) = Just C0
+  decodeTerm (CBOR.TInt 1) = Just C1
+  decodeTerm (CBOR.TInt 2) = Just C2
+  decodeTerm _             = Nothing
+
+instance Arbitrary Data_0 where
+  arbitrary = oneof [return C0, return C1, return C2]
+  shrink C2 = [C1, C0]
+  shrink C1 = [C0]
+  shrink C0 = []
+
+instance CoArbitrary Data_0 where
+
+-- |
+-- Data associated with @'Version_1'@
+--
+data Data_1 = Data_1 Bool
+  deriving (Eq, Show, Typeable, Generic)
+
+instance SerialiseTerm Data_1 where
+  encodeTerm (Data_1 b) = CBOR.TBool b
+
+  decodeTerm (CBOR.TBool b) = Just (Data_1 b)
+  decodeTerm _              = Nothing
+
+instance Arbitrary Data_1 where
+  arbitrary = Data_1 <$> arbitrary
+  shrink (Data_1 n) = map Data_1 (shrink n)
+
+instance CoArbitrary Data_1
+
+-- |
+-- Data associated with @'Version_2'@
+--
+data Data_2 = Data_2 Word Word
+  deriving (Eq, Show, Typeable, Generic)
+
+instance SerialiseTerm Data_2 where
+  encodeTerm (Data_2 n m) = CBOR.TList [CBOR.TInt (fromIntegral n), CBOR.TInt (fromIntegral m)]
+  decodeTerm (CBOR.TList (CBOR.TInt n : CBOR.TInt m : _)) = Just (Data_2 (fromIntegral n) (fromIntegral m))
+  decodeTerm _ = Nothing
+
+instance Arbitrary Data_2 where
+  arbitrary = Data_2 <$> arbitrary <*> arbitrary
+  shrink (Data_2 n m) =
+    [ Data_2 n m'
+    | m' <- shrink m
+    ]
+    ++
+    [ Data_2 n' m
+    | n' <- shrink n
+    ]
```

Comment:
Simple shrinkers like this can use `shrinkGeneric`, or can reuse the one for `(,)`. Saves a bit of typing.